cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.queue;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.id.ScheduleId;
+
+/**
+ * Responsible for keeping track of {@link Job}s, which correspond to schedules that have been triggered,
+ * but not yet executed.
+ */
+interface JobQueue {
+
+  /**
+   * Returns a {@link CloseableIterator} over all the jobs associated with the given schedule Id.
+   */
+  CloseableIterator<Job> getJobsForSchedule(ScheduleId scheduleId);
+
+  /**
+   * Returns a stored Job, given the scheduleId and the creationTime of it.
+   *
+   * @param jobKey the key for the requested Job
+   * @return the stored Job.
+   */
+  Job getJob(JobKey jobKey);
+
+  /**
+   * Creates a new Job in the queue or updates an existing Job.
+   *
+   * @param job the new job
+   */
+  void put(Job job);
+
+  /**
+   * Updates a given Job to have a new {@link Job.State}.
+   *
+   * @param state the new job state
+   */
+  void updateState(Job job, Job.State state);
```

Comment:
From the implementation, this method might throw `IllegalArgumentException` is the state provided is not valid based on the state transition. Please add that to javadoc.